document user-facing functions in utils

otizonaizit opened the pull request:
Document user-facing functions

Fixes: #671

otizonaizit commented:
@Debilski : ready for review!

Debilski commented on pelita/game.py:
```diff
@@ -81,18 +81,101 @@ def controller_exit(state, await_action='play_step'):
+              See below in section Notes for more details about local and remote games.
+
+    layout_dict : dict
+               a dictionary representing a maze, as returned by pelita.layout.parse_layout
```
Space below.

Debilski commented on pelita/game.py:
```diff
@@ -81,18 +81,101 @@ def controller_exit(state, await_action='play_step'):
+                  client) to return. After timeout_length seconds are elapsed an
+                  non-fatal error is recorded for the team.
+
+    viewers : list[viewer1, viewer2]
```
`list[viewer1, viewer2, ...]` to avoid implying there have to be two viewers (as with the teams)

Debilski commented on pelita/game.py:
```diff
@@ -81,18 +81,101 @@ def controller_exit(state, await_action='play_step'):
+                    allow_exceptions should be False, so that the game master can
+                    collect the exceptions and cleanly create a game-over state if
+                    needed.
+
```
Indendation is a little inconsistent. Either one or two characters left of the : 
Maybe ignore the : and just make it 4 everywhere?

Debilski commented on pelita/game.py:
```diff
@@ -81,18 +81,101 @@ def controller_exit(state, await_action='play_step'):
+               - the path to a Python module that defines at least a function
+                 called 'move' (specified as above), and a string TEAM_NAME.
+
+              When the team_specs are functions a "local" game will be run, when
```
I’d prefer: ‘When a team_spec is a function’ because they can be different.

Debilski replied:
+ comma after when clause, no?

Debilski commented on pelita/game.py:
```diff
@@ -81,18 +81,101 @@ def controller_exit(state, await_action='play_step'):
+                 position (x, y) of bot and state.
+                 state can be an arbitrary Python object, None by default
+               - the path to a Python module that defines at least a function
+                 called 'move' (specified as above), and a string TEAM_NAME.
```
No comma unless you put one before ‘that’

Debilski commented on pelita/game.py:
```diff
@@ -81,18 +81,101 @@ def controller_exit(state, await_action='play_step'):
+                 called 'move' (specified as above), and a string TEAM_NAME.
+
+              When the team_specs are functions a "local" game will be run, when
+              they are paths to Python modules a "remote" game will be run instead.
```
… paths to Python modules _ comma _  a "remote" game will be run instead

Debilski commented on pelita/game.py:
```diff
@@ -81,18 +81,101 @@ def controller_exit(state, await_action='play_step'):
+                    when True, allow teams to raise Exceptions. This is especially
+                    useful when running local games, where you typically want to
+                    see Exceptions do debug them. When running remote games,
+                    allow_exceptions should be False, so that the game master can
```
No game master.

Debilski commented on pelita/game.py:
```diff
@@ -81,18 +81,101 @@ def controller_exit(state, await_action='play_step'):
+        If teams_specs is a list of two Python modules, a "remote" game will be
+        played. This means that the game master will create one client subprocess for
+        each module. The game master communicates with the clients using the local
+        network and sends/receives messages in JSON format. The game master will ask
```
No game master

Debilski commented on pelita/scripts/pelita_main.py:
```diff
@@ -12,7 +12,10 @@
 import pelita
-from pelita import game, layout, utils
+from pelita import game, layout
+
```
I think the other imports could also be made relative.

Debilski commented on pelita/utils.py:
```diff
@@ -1,36 +1,33 @@
+    Parameters
+    ----------
+    walls : list[(x0,y0), (x1,y1), ...]
+         a list of walls coordinates
```
wall coordinates

Debilski commented on pelita/utils.py:
```diff
@@ -1,36 +1,33 @@
+
+    Notes
+    -----
+    Nodes in the graph are (x,y) coordinates representing a square in the maze
```
squares

Debilski commented on pelita/utils.py:
```diff
@@ -55,21 +52,74 @@ def load_builtin_layout(layout_name, *, is_blue=True):
+                     '''
+
+    is_blue : bool
+           when True setups up the game assuming your bots are in the blue team,
```
when True, sets up the game

Debilski commented on pelita/utils.py:
```diff
@@ -55,21 +52,74 @@ def load_builtin_layout(layout_name, *, is_blue=True):
+    is_blue : bool
+           when True setups up the game assuming your bots are in the blue team,
+           i.e. the team with the homezone on the left side of the maze. Otherwise
+           your bots are assumed to be in the red team.
```
I think ‘Otherwise’ is misleading here as it implies that this also happens when you leave it out. But the default is True. So, I’d just write ‘when False’.

Debilski commented on pelita/utils.py:
```diff
@@ -55,21 +52,74 @@ def load_builtin_layout(layout_name, *, is_blue=True):
+    bots : list[(x0,y0), (x1,y1)]
+        list of coordinates for your bots. These will override the positions found
+        in the layout. If only one pair of coordinates is given, i.e. you pass just
+        [(x0,y0)] only the position for bot '0' will be set
```
, only the position for bot '0' will be overridden

Debilski commented on pelita/utils.py:
```diff
@@ -55,21 +52,74 @@ def load_builtin_layout(layout_name, *, is_blue=True):
+         pass just [(x0,y0)] only the position for enemy bot '0' will be set
+
+    is_noisy : list[bool, bool]
+            list of two booleans for the enemy bots is_noisy property
```
the enemy bots’

Debilski commented on pelita/utils.py:
```diff
@@ -55,21 +52,74 @@ def load_builtin_layout(layout_name, *, is_blue=True):
     layout = create_layout(layout, food=food, bots=bots, enemy=enemy, is_noisy=is_noisy)
     width = max(layout['walls'])[0] + 1
 
+    def split_food(width, food):
```
We could also reuse the one in game

Debilski commented:
Can the background game go in a separate PR? This is quite unrelated.

otizonaizit commented:
>Can the background game go in a separate PR? This is quite unrelated.

It could, but I am out of steam for today. Given that it is indeed just adding a user-facing function, the documentation of which is the core part, it would be nice if you just reviewed this too right here. I will do the changes from your comments tomorrow. Or, if you have time, you can do them yourself, I think the PR is open to changes from maintainers?

Debilski commented:
I think you have to explicitly allow that changes can be added

Debilski commented:
It is a user-facing function but I do have more ideas for it that deserve a separate discussion.

otizonaizit commented:
> I think you have to explicitly allow that changes can be added

it's enabled

otizonaizit commented:
> It is a user-facing function but I do have more ideas for it that deserve a separate discussion.

can we merge this first, and then discuss whatever you want when you have an alternative or improved implementation, together with a PR on pelita_template which uses your improved version? This would help in cutting the theoretical discussions and focusing on the concrete lines of code.

Debilski commented:
In its current form, I am not sure how useful the background game function is. There is no way to look at the game or replay it so all the data it collects is not super meaningful.

otizonaizit commented:
What date

>In its current form, I am not sure how useful the background game
>function is. There is no way to look at the game or replay it so all
>the data it collects is not super meaningful.

What are you talking about? Look at the example usage in the PR on pelita_template. You have everything you need, and in the example I give there it's pretty clear how useful it is. Playing games on the background is done to collect stats.
 If you want I can add there an example where using the seed and the layout name, which are both stored, you replay a game in the TK UI. Is this what you mean?

otizonaizit commented:
>Why did these go? I think they are still useful together with
>`setup_test_game`.

How? I can adapt setup_test_game to use the same layout logic of run_background_name and we can also get rid of create_layout, which is now redundant...

Debilski commented:
> Is this what you mean?

https://github.com/ASPP/pelita/pull/676

Debilski commented:
But an example how to replay with a seed would be nice, too. I am not really sure if it works. In my tests it didn’t. I hope it is fixable.

otizonaizit commented:
I think I have addressed your concerns expressed in #676 with https://github.com/ASPP/pelita_template/commit/1e7ec326b8c889bc04a3744ddf2911b561c17df0 and https://github.com/ASPP/pelita/commit/6172522949862fbae7510824b92817fd77a4185d .

I am asusming this branch will be merged very soon so that I can go and address https://github.com/ASPP/pelita_template/issues/23 , which is in my opinion the last road-block to ASPP 2019. 

I am opening an issue for adding tests to `run_background_game` and another one for the circular imports in `utils.py`.
